Add Ledger key rotation tutorial

alnoki opened the pull request:
# Background

@davidiw @gedigi @hariria  @hardsetting @xbtmatt 

Per in-person discussions re: authentication key rotation and Ledger.

This relies on the new features from the `aptos-core` PR https://github.com/aptos-labs/aptos-core/pull/11151 which was subsequently broken into 3 smaller PRs:

- https://github.com/aptos-labs/aptos-core/pull/14084
- https://github.com/aptos-labs/aptos-core/pull/14266
- https://github.com/aptos-labs/aptos-core/pull/14309

# Changes

1. Add a tutorial on rotating the authentication key for a wallet from hot private key to Ledger and back, using localnet to demonstrate new Framework code.
2. Change a fenced code block with lots of escape characters from `bash` to `text` to avoid syntax highlighting issues in IDE.
3. Update existing authentication key rotation documentation as necessary with important caveats.
4. Crosslink Ledger and standard key rotation documents.
5. Fix header hierarchy where it resulted in unintuitive navigation.

# Testing

From in `apps/nextra`:

```sh
npx -p pnpm@8.9.0 pnpm dev
```

1. Install the `aptos-core` CLI [from source] using https://github.com/aptos-labs/aptos-core/pull/14309 (or `main`, once it merges)
1. Follow http://localhost:3030/en/build/guides/key-rotation
1. Follow http://localhost:3030/en/build/cli/trying-things-on-chain/ledger#authentication-key-rotation

# Checklist

- Do all Lints pass?
  - [x] Have you ran `pnpm spellcheck`? (This is failing due to issues with content I have not modified)
  - [x] Have you ran `pnpm fmt`?
  - [x] Have you ran `pnpm lint`? (This is failing due to issues with content I have not modified)

[from source]: https://aptos.dev/en/build/cli/install-cli/install-cli-specific-version

gedigi commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
```diff
@@ -117,3 +117,266 @@ yes
+
+### Set up localnet hot wallet profile
+
+In a fresh terminal window, set up a localnet profile for an account with a compromised private key:
```
it's not clear why I need to set a compromised key (what does compromised mean in this case? leaked?) and where the value below comes from.

alnoki replied:
Removed as of 6b6cb8c

gregnazario commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
```diff
@@ -117,3 +117,266 @@ yes
+
+### Set up localnet hot wallet profile
+
+In a fresh terminal window, set up a localnet profile for an account with a compromised private key:
```
I wouldn't make this a compromised key, I was just create a new profile with a random key for the purposes of the tutorial.  Either that or just say with a "fixed private key" for the purposes of the tutorial.

Compromised shouldn't really come into play here.  The only reason it is compromised is because it's hardcoded a paragraph down.

alnoki replied:
Removed as of 6b6cb8c

gregnazario commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
```diff
@@ -117,3 +117,266 @@ yes
+Open up a new terminal window to continue with the rest of the tutorial and leave the localnet running in the background.
+
+<Callout type="default">
+When you finally want to shut down the localnet, you can press `Ctrl+C` in the terminal window where you started the localnet.
```
Does this work in Windows?

alnoki replied:
Addressed as of 6b6cb8c

gregnazario commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
````diff
@@ -117,3 +117,266 @@ yes
+This tutorial will walk you through both scenarios.
+
+
+<Callout type="default" emoji="🧠">
+Before you start this tutorial make sure you have `jq`, a lightweight and flexible command-line JSON processor:
+
+```sh filename="Terminal"
+jq --version
+```
+</Callout>
+
+<Steps>
````
`jq` is specific to `unix`, so for the purpose of this tutorial, it should say something that rotation is supported in Windows, but this tutorial is for macOS / Linux

alnoki replied:
Addressed as of 6b6cb8c

gregnazario commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
````diff
@@ -117,3 +117,266 @@ yes
+
+Rotate the authentication key of the hot wallet via [`aptos_framework::account::rotate_authentication_key_call`](https://github.com/aptos-labs/aptos-core/blob/415d7e12e94ebd833a910ed4138742f005149610/aptos-move/framework/aptos-framework/sources/account.move#L294):
+
+```sh filename="Terminal"
+aptos move run \
+    --args "hex:$AUTH_KEY_B" \
+    --profile localnet-a \
+    --function-id 0x1::account::rotate_authentication_key_call
+```
````
Should probably just change the existing rotate key to use this then, and then you can get rid of most of this tutorial?

alnoki replied:
Addressed as of 6b6cb8c, combined with https://github.com/aptos-labs/aptos-core/pull/11151

alnoki commented:
Per @gregnazario in https://github.com/aptos-labs/developer-docs/pull/367#discussion_r1623092630:

> Should probably just change the existing rotate key to use this then, and then you can get rid of most of this tutorial?

I had previously tried to incorporate Ledger rotation support in the CLI https://github.com/aptos-labs/aptos-core/pull/11151 and got blocked by Ledger's inability to sign the `RotationProofChallenge`, which is why I was using the new `rotate_authentication_key_call` here. (Unfortunately, however, the method presented in this PR doesn't update the `OriginatingAddress` table).

However I did some digging and it now looks like it now may be possible to sign the `RotationProofChallenge` with Ledger.

Hence I've created https://github.com/aptos-labs/aptos-core/issues/13515 to track the CLI feature, which I agree would be more straightforward cc @gedigi

gregnazario commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
````diff
@@ -117,3 +119,495 @@ yes
+If you are using a UNIX-like machine:
+
+```shell filename="Terminal"
+aptos config delete-profile --profile ledger-wallet-1000
````
This is also not a command?

alnoki replied:
As I mentioned in the PR description

> Install the CLI from source, as described in https://github.com/aptos-labs/aptos-core/pull/11151

This PR pairs simply because I had to split docs and source into two separate repos https://github.com/aptos-labs/aptos-core/pull/11151

alnoki replied:
Noting that I've updated the PR description per the recent set of `aptos-core` PRs that have merged, to stipulate installing the latest Framework updates

gregnazario reviewed (commented):
@movekevin can you take a look at this?

alnoki commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
```diff
@@ -117,3 +119,495 @@ yes
+
+This tutorial will walk you through both scenarios.
+
+<Callout type="warning" emoji="💀">
```
Addressed in 543aaa6

Note that the callout additionally specifies not to do it until finishing the base ledger guide, which includes a power user warning

alnoki commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
```diff
@@ -117,3 +119,495 @@ yes
+
+<Callout type="info" emoji="🧠">
+If you are on a UNIX-like system, the following command can be used to start a
+fresh localnet as a background process:
```
Addressed in 543aaa6 with MacOS/Linux distinction and power user callout

alnoki commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
````diff
@@ -117,3 +119,495 @@ yes
+
+### Set up localnet hot wallet profile
+
+Create a private key corresponding to an authentication key, and thus initial
+account address, that starts with the vanity prefix `0xaaa`:
+
+```sh filename="Terminal"
+aptos key generate \
+    --assume-yes \
+    --output-file private-key-a \
+    --vanity-prefix 0xaaa
+```
````
It can actually be kinda hard to follow along if luck of the draw two accounts start with similar hex chars

Vanity address with only 3 chars is negligible time to generate, and the labelling make the tutorial flow way easier

alnoki commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
```diff
@@ -117,3 +119,495 @@ yes
+Ledger.
+
+This practice aids in profile recovery, as shown below.
+</Callout>
```
Addressed in 543aaa6

alnoki commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
````diff
@@ -117,3 +119,495 @@ yes
+```sh filename="Terminal"
+aptos account rotate-key \
+    --assume-yes \
+    --new-derivation-index 1000 \
````
As I mentioned in the PR description

> Install the CLI from source, as described in https://github.com/aptos-labs/aptos-core/pull/11151

This PR pairs simply because I *had* to split docs and source into two separate repos https://github.com/aptos-labs/aptos-core/pull/11151

alnoki replied:
Noting that I've updated the PR description per the recent set of `aptos-core` PRs that have merged, to stipulate installing the latest Framework updates

alnoki commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
````diff
@@ -117,3 +119,495 @@ yes
+```shell filename="Terminal"
+rm private-key-*
+```
+</Callout>
````
Addressed in https://github.com/aptos-labs/developer-docs/commit/543aaa6836e063fb3c645aedce01b5c94fb85583

alnoki commented:
@gregnazario thanks for the above comments. I believe I've addressed all where applicable

Just flagging here again that this PR requires https://github.com/aptos-labs/aptos-core/pull/11151, without which there is no support for ledger auth key rotation

I'd try and do everything in one PR (docs and CLI source updates), but alas there is no longer a monorepo

hariria commented on apps/nextra/pages/en/build/guides/key-rotation.mdx:
```diff
@@ -30,97 +31,773 @@ Here are the installation links for the SDKs we will cover in this example:
+users to query an account address associated with a given authentication key,
+and only allows for one entry per authentication key. Hence the requirement of a
+signed [`account::RotationProofChallenge`] to ensure that a malicious actor does
+not rotate an account's authentication key to a key that is already in the
+table, as this attack that would prevent lookup of the valid originating address
+that the holder of an authentication key had previously approved.
```
```suggestion
not rotate an account's authentication key to a key that is already in the
table, as this attack would prevent lookup of the valid originating address
that the holder of an authentication key had previously approved.
```

alnoki replied:
Addressed in 806a984

hariria commented on apps/nextra/pages/en/build/guides/key-rotation.mdx:
```diff
@@ -30,97 +31,773 @@ Here are the installation links for the SDKs we will cover in this example:
+rotation while the associated authentication key is mapped to a second arbitrary
+account address in the [`account::OriginatingAddress`] table, because this
+operation would fail the check that the rotating account's address is the one
+mapped to in the table (since the table is only updated upon during rotation,
+not upon standard account generation).
```
```suggestion
mapped to in the table (since the table is only updated during rotation,
not upon standard account generation).
```

alnoki replied:
Addressed in 806a984

hariria commented on apps/nextra/pages/en/build/guides/key-rotation.mdx:
```diff
@@ -30,97 +31,773 @@ Here are the installation links for the SDKs we will cover in this example:
+
+<Callout type="info" emoji="🧠">
+The [`account::rotate_authentication_key_call`] was introduced to support
+non-standard key algorith, like passkeys, which cannot produce proofs of
```
```suggestion
non-standard key algorithms, like passkeys, which cannot produce proofs of
```

alnoki replied:
Addressed in 806a984

hariria commented on apps/nextra/pages/en/build/guides/key-rotation.mdx:
```diff
@@ -30,97 +31,773 @@ Here are the installation links for the SDKs we will cover in this example:
+if so, verifies that the rotating account's address is the one mapped to in the
+table.
+
+This means that if an arbitary account's authentication key is rotated to
```
```suggestion
This means that if an arbitrary account's authentication key is rotated to
```

alnoki replied:
Addressed in 806a984

hariria reviewed (approved):
Some nits and otherwise it looks good to me. Will let @gregnazario take a final pass

hariria commented on apps/nextra/pages/en/build/cli/trying-things-on-chain/ledger.mdx:
```diff
@@ -117,3 +119,499 @@ yes
+
+<Callout type="warning" emoji="❗">
+Before you start this tutorial make sure you have completed the
+[key rotation guide](../../advanced-guides/key-rotation.mdx).
```
All of the relative links to the key rotation guide do not work for me when testing it in the preview.

alnoki replied:
It looks like `advanced-guides` was renamed to `guides`

Resolved in c32f593

alnoki commented:
@hariria I believe I've addressed all your recent comments and that I had previously addressed all comments from @gregnazario 

I've updated the PR description for more clarity on the associated `aptos-core` PRs

igor-p commented:
Discussed with @gregnazario, ok to merge.
